fem/fe/fe_rbf.hpp

```diff
@@ -0,0 +1,556 @@
+   }
+};
+
+// Dimensionless distance metrics, whose output is the input for RBFs
+class DistanceMetric
+{
+protected:
+   int dim;
+public:
+   DistanceMetric(int D) { dim = D; }
+   virtual ~DistanceMetric() { }
+
+   virtual void SetDim(int D) { dim = D; }
+
+   virtual void Distance(const Vector &x,
+                         double &r) const = 0;
+   virtual void DDistance(const Vector &x,
+                          Vector &dr) const = 0;
+   virtual void DDDistance(const Vector &x,
+                           DenseMatrix &ddr) const = 0;
+
+   static DistanceMetric *GetDistance(int dim, int pnorm);
+};
+
+// Dimensionless distance with r = |x| + |y| + ...
+class L1Distance : public DistanceMetric
+{
+public:
+   L1Distance(int D) : DistanceMetric(D) { };
+   virtual ~L1Distance() { }
+
+   virtual void Distance(const Vector &x,
+                         double &r) const;
+   virtual void DDistance(const Vector &x,
+                          Vector &dr) const;
+   virtual void DDDistance(const Vector &x,
+                           DenseMatrix &ddr) const;
+};
+
+// Dimensionless distance with r = (x^2 + y^2 + ...)^(1/2)
+class L2Distance : public DistanceMetric
+{
+public:
+   L2Distance(int D) : DistanceMetric(D) { };
+   virtual ~L2Distance() { }
+
+   virtual void Distance(const Vector &x,
+                         double &r) const;
+   virtual void DDistance(const Vector &x,
+                          Vector &dr) const;
+   virtual void DDDistance(const Vector &x,
+                           DenseMatrix &ddr) const;
+};
+
+// Dimensionless distance with r = (x^p + y^p + ...)^(1/p)
+class LpDistance : public DistanceMetric
+{
+   const int p;
+   const double pinv;
+public:
+   LpDistance(int D, int pnorm)
+      : DistanceMetric(D),
+        p(pnorm),
+        pinv(1. / static_cast<double>(p))
+   { };
+   virtual ~LpDistance() { }
+
+   virtual void Distance(const Vector &x,
+                         double &r) const;
+   virtual void DDistance(const Vector &x,
+                          Vector &dr) const;
+   virtual void DDDistance(const Vector &x,
+                           DenseMatrix &ddr) const;
+};
+
+
+/* Pure virtual class for a finite element with radial basis functions
+   instead of polynomials inside each element */
+class KernelFiniteElement : public ScalarFiniteElement
+{
+private:
+   // Choose whether to interpolate or project when Project is called
+   bool interpolate = false;
+public:
+   KernelFiniteElement(int D, Geometry::Type G, int Do, int O, int F)
+      : ScalarFiniteElement(D, G, Do, O, F) { }
+   virtual ~KernelFiniteElement() { }
+
+   // Converts integration rule to vector
+   virtual void IntRuleToVec(const IntegrationPoint &ip,
+                             Vector &vec) const;
+
+   // Is base RBF compact?
+   virtual bool IsCompact() const = 0;
+
+   // Return base kernel
+   virtual const RBFFunction *Kernel() const = 0;
+
+   // Get range of i,j,k indices that are nonzero for compact support
```

Comment:
Could you explain what `indices` means?

Reply:
Done. See https://github.com/mfem/mfem/blob/meshless-kernels/fem/fe/fe_rbf.hpp#L376-L388.